tests/Tests/Unit/ResourceHelperTraitTest.php

```diff
@@ -0,0 +1,115 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+namespace Google\ApiCore\Tests\Unit;
+
+use Google\ApiCore\PathTemplate;
+use Google\ApiCore\ResourceHelperTrait;
+use Google\ApiCore\ValidationException;
+use PHPUnit\Framework\TestCase;
+use Yoast\PHPUnitPolyfills\Polyfills\ExpectException;
+
+class ResourceHelperTraitTest extends TestCase
+{
+    use ExpectException;
+
+    public function testLoadTemplates()
+    {
+        $got = ResourceHelperTraitStub::testLoadTemplates();
+        $this->assertEquals(count($got), 4);
+        $this->assertTrue($got['project'] instanceof PathTemplate);
+    }
+
+    public function testGetTemplate()
+    {
+        $got = ResourceHelperTraitStub::testGetTemplate('project');
+        $this->assertNotNull($got);
+        $this->assertTrue($got instanceof PathTemplate);
+    }
+
+    public function testGetTemplateNull()
+    {
+        $got = ResourceHelperTraitStub::testGetTemplate('does_not_exist');
+        $this->assertNull($got);
+    }
+
+    public function testParseName()
+    {
+        $got = ResourceHelperTraitStub::parseName('projects/abc123', 'project');
+        $this->assertEquals(count($got), 1);
+        $this->assertEquals($got['project'], 'abc123');
+    }
+
+    public function testParseNameInvalidTemplate()
+    {
+        $this->expectException(ValidationException::class);
+        $this->expectExceptionMessage('Template name does_not_exist does not exist');
+
+        ResourceHelperTraitStub::parseName('projects/abc123', 'does_not_exist');
+    }
+
+    public function testParseNameNoMatchingPattern()
+    {
+        $this->expectException(ValidationException::class);
+        $this->expectExceptionMessage('Input did not match any known format. Input: no/matching/pattern');
+
+        ResourceHelperTraitStub::parseName('no/matching/pattern');
+    }
+    
+}
+
+class ResourceHelperTraitStub{
+    use ResourceHelperTrait;
+
+    public static $CONFIG_PATH = __DIR__ . '/testdata/test_service_descriptor_config.php';
+    public static $SERVICE_NAME = 'test.interface.v1.api';
```

Comment:
I'm going to guess this was done in case we wanted to customize it to point to a different file or service name in another test. I'd rather it actually mimic the GAPICs though! (which means, using constants but not private ones...)

Reply:
For sure, sounds good. Done.